editor/components/block-drop-zone/index.js

```diff
@@ -14,51 +14,88 @@ import {
 	getBlockTransforms,
 	findTransform,
 } from '@wordpress/blocks';
-import { compose } from '@wordpress/element';
+import { compose, Component } from '@wordpress/element';
 
 /**
  * Internal dependencies
  */
-import { insertBlocks, updateBlockAttributes } from '../../store/actions';
+import { insertBlocks, updateBlockAttributes, moveBlockToPosition } from '../../store/actions';
 
-function BlockDropZone( { index, isLocked, ...props } ) {
-	if ( isLocked ) {
-		return null;
+class BlockDropZone extends Component {
+	constructor() {
+		super();
+
+		this.onDropFiles = this.onDropFiles.bind( this );
+		this.onHTMLDrop = this.onHTMLDrop.bind( this );
+		this.onDrop = this.onDrop.bind( this );
 	}
 
-	const getInsertIndex = ( position ) => {
+	getInsertIndex( position ) {
+		const { index } = this.props;
 		if ( index !== undefined ) {
 			return position.y === 'top' ? index : index + 1;
 		}
-	};
+	}
 
-	const onDropFiles = ( files, position ) => {
+	onDropFiles( files, position ) {
 		const transformation = findTransform(
 			getBlockTransforms( 'from' ),
 			( transform ) => transform.type === 'files' && transform.isMatch( files )
 		);
 
 		if ( transformation ) {
-			const insertIndex = getInsertIndex( position );
-			const blocks = transformation.transform( files, props.updateBlockAttributes );
-			props.insertBlocks( blocks, insertIndex );
+			const insertIndex = this.getInsertIndex( position );
+			const blocks = transformation.transform( files, this.props.updateBlockAttributes );
+			this.props.insertBlocks( blocks, insertIndex );
 		}
-	};
+	}
 
-	const onHTMLDrop = ( HTML, position ) => {
+	onHTMLDrop( HTML, position ) {
 		const blocks = rawHandler( { HTML, mode: 'BLOCKS' } );
 
 		if ( blocks.length ) {
-			props.insertBlocks( blocks, getInsertIndex( position ) );
+			this.props.insertBlocks( blocks, this.getInsertIndex( position ) );
+		}
+	}
+
+	onDrop( event, position ) {
+		if ( ! event.dataTransfer ) {
+			return;
+		}
+
+		let uid, type, rootUID, fromIndex;
+
+		try {
+			( { uid, type, rootUID, fromIndex } = JSON.parse( event.dataTransfer.getData( 'text' ) ) );
+		} catch ( err ) {
+			return;
+		}
+
+		if ( type !== 'block' ) {
+			return;
 		}
-	};
-
-	return (
-		<DropZone
-			onFilesDrop={ onDropFiles }
-			onHTMLDrop={ onHTMLDrop }
-		/>
-	);
+		const { index } = this.props;
+		const positionIndex = this.getInsertIndex( position );
+		// If the block is kept at the same level and moved downwards,
+		// we need to substract "1" from the insert index.
```

Comment:
My first though was "but why?" After thinking some more, I started to understand it's to account for the fact that the block is no longer where it was, so all other blocks below it shift upward. Could be more clear.

Reply:
I don't know how to express this honestly :) Some help would be appreciated.

Reply:
I had a hard time explaining what happens when you remove an item from a list and insert it in a different position. The indexes change. Maybe my response here can shed some light: https://github.com/WordPress/gutenberg/pull/4115#discussion_r163889376 (about creating a general function to move 1+ blocks to a different index). If not, give me a ping to try again.

Reply:
Maybe:

```js
// If the block is kept at the same level and moved downwards, subtract
// to account for blocks shifting upward to occupy its old position.
```